## Final Project Submission

Please fill out:
* Student name: Dominic Muthiani Muli
* Student pace: part time
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


# Movie Genre Analysis For Proposed Microsoft Studio

<img src="./images/vecteezy_microsoft-logo-on-transparent-background_14018578.jpg" alt="Microsoft Logo" width="300">

## 1.0) Project Overview

[Microsoft Corporation](https://www.microsoft.com/en-us/) is a multinational computer technology corporation that develops, licenses, and sells computer software, consumer electronics, and personal computers. It is headquartered in Redmond, Washington. The company was founded on April 4, 1975, by Bill Gates and Paul Allen in Albuquerque, New Mexico.

This project analyses movie genre ratings and gross earnings to offer Microsoft Corporation insights about movies for the proposed new original movie studio.

## 1.1) Business Problem

<img src="./images/Clapper_board_Clapper-board.jpg" alt="Movie" height="200">

Microsoft plans to venture in original video content creation. The head of Microsoft's proposed movie studio wants to know what genres of films are currently performing best at the box office. They need this information so that they can focus on creating films that are likely to succeed.

## 1.2) Objectives

* Identify the best performing movie genres at the box office.
* Generate recommendations for the types of films that Microsoft's proposed movie studio should create.

## 1.3) Metric of Success

The project will be considered successful if:
* I accurately make predictions about which movie genres are doing the best at the box office
* I generate useful recommendations about types of films the studio should create

## 1.4) Data Sources

The data sources for this data science project will be:

* The bom.movie_gross.csv dataset, which contains information about the gross earnings of movies from [Box Office Mojo](https://www.boxofficemojo.com/).
* The title.basics.csv dataset, which contains information about the basic details of movies from [IMDB](https://www.imdb.com/).
* The title.ratings.csv dataset, which contains information about the ratings of movies from [IMDB](https://www.imdb.com/).


## 1.5) Data Relevance and Validation

The data available is relevant for the intended analysis. More information can be found here:
*  [IMDB](https://www.imdb.com/)
*  [Box Office Mojo](https://www.boxofficemojo.com/)

## 2.0) Reading and Understanding the Data

In [1]:
# importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
pd.set_option('display.max_columns', None)

In [2]:
# reading the data
movie_gross = pd.read_csv("./data/bom.movie_gross.csv")
movie_info = pd.read_csv("./data/title.basics.csv")
movie_ratings = pd.read_csv("./data/title.ratings.csv")

## 2.1) Understanding the Data

In [3]:
movie_gross.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


In [4]:
movie_gross.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                   Alice in Wonderland (2010)     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

  foreign_gross  year  
0     652000000  2010  
1     691300000  2010  
2     664300000  2010  
3     535700000  2010  
4     513900000  2010

In [5]:
movie_gross["year"].value_counts().sort_index()

2010    328
2011    399
2012    400
2013    350
2014    395
2015    450
2016    436
2017    321
2018    308
Name: year, dtype: int64

**Movies gross earnings** dataset has 3387 records and 5 columns. The data shows different movie titles, studio they were recorded from, revenue (domestic and foreign) and year. The data runs from 2010 through 2018.

In [6]:
movie_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146144 entries, 0 to 146143
Data columns (total 6 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   tconst           146144 non-null  object 
 1   primary_title    146144 non-null  object 
 2   original_title   146123 non-null  object 
 3   start_year       146144 non-null  int64  
 4   runtime_minutes  114405 non-null  float64
 5   genres           140736 non-null  object 
dtypes: float64(1), int64(1), object(4)
memory usage: 6.7+ MB


In [7]:
movie_info.head()

tconst                    primary_title              original_title  \
0  tt0063540                        Sunghursh                   Sunghursh   
1  tt0066787  One Day Before the Rainy Season             Ashad Ka Ek Din   
2  tt0069049       The Other Side of the Wind  The Other Side of the Wind   
3  tt0069204                  Sabse Bada Sukh             Sabse Bada Sukh   
4  tt0100275         The Wandering Soap Opera       La Telenovela Errante   

   start_year  runtime_minutes                genres  
0        2013            175.0    Action,Crime,Drama  
1        2019            114.0       Biography,Drama  
2        2018            122.0                 Drama  
3        2018              NaN          Comedy,Drama  
4        2017             80.0  Comedy,Drama,Fantasy

In [8]:
# Comparing movie_info primary_title and original_title
movie_info[movie_info['primary_title'] != movie_info['original_title']][['primary_title', 'original_title']]

primary_title  \
1          One Day Before the Rainy Season   
4                 The Wandering Soap Opera   
11                    So Much for Justice!   
13            Children of the Green Dragon   
15                      The Tragedy of Man   
...                                    ...   
146026          Journey of the Sky Goddess   
146028  Lupin the Third: Fujiko Mine's Lie   
146037                   Big Three Dragons   
146121                       A Cherry Tale   
146135                       The Rehearsal   

                             original_title  
1                           Ashad Ka Ek Din  
4                     La Telenovela Errante  
11                           Oda az igazság  
13                 A zöld sárkány gyermekei  
15                      Az ember tragédiája  
...                                     ...  
146026                  Kibaiyanse! Watashi  
146028  Lupin the IIIrd: Mine Fujiko no Uso  
146037                          Da San Yuan  
146121                    Kirsebæreventyret  
146135                             O Ensaio  

[14504 rows x 2 columns]

In [9]:
# understanding title column of movie_gross
movie_gross[movie_gross["title"]=="O Ensaio"]

Empty DataFrame
Columns: [title, studio, domestic_gross, foreign_gross, year]
Index: []

**Movies information** dataset has 146144 records and 6 columns. The data shows different movie primary title in english, original titles in origninal language, year the movie started airing, movie run time in minutes and genres the movie belongs to.

In [10]:
movie_ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   tconst         73856 non-null  object 
 1   averagerating  73856 non-null  float64
 2   numvotes       73856 non-null  int64  
dtypes: float64(1), int64(1), object(1)
memory usage: 1.7+ MB


In [11]:
movie_ratings.head()

tconst  averagerating  numvotes
0  tt10356526            8.3        31
1  tt10384606            8.9       559
2   tt1042974            6.4        20
3   tt1043726            4.2     50352
4   tt1060240            6.5        21

**Movies ratings** dataset has 73856 records and 36 columns. The data shows different movie identifiers, movie weighted average user rating on a scale of 1 to 10, and number of votes.

## 2.2) Preparing the Data

### 2.2.1) Data Cleaning

movie_gross foreign is formated as sting intead of a float

In [12]:
# A function to identify non numeric rows in "foreign_gross"
def non_numeric_rows(data):
  non_numeric_df = movie_gross[pd.to_numeric(data, errors='coerce').isna()]
  return non_numeric_df

In [13]:
non_numeric_foreign_gross = non_numeric_rows(movie_gross["foreign_gross"])
non_numeric_foreign_gross.sample(10)

title  studio  domestic_gross  \
1029                             Love in the Buff      CL        256000.0   
1794                                    Bethlehem      AF        202000.0   
2211            Ghost in the Shell: The New Movie     FUN        101000.0   
2946                                     Colossal    Neon       3000000.0   
2055                                      Dilwale     UTV       4900000.0   
2959          Along with the Gods: The Two Worlds   WGUSA       1900000.0   
2573                           Operation Chromite      CJ       1000000.0   
3376  Lou Andreas-Salomé, the Audacity to be Free     CLS         12900.0   
2715                               Valley of Love  Strand         28600.0   
2254                                    Wild City   WGUSA         34900.0   

     foreign_gross  year  
1029           NaN  2012  
1794           NaN  2014  
2211           NaN  2015  
2946           NaN  2017  
2055           NaN  2015  
2959           NaN  2017  
2573           NaN  2016  
3376           NaN  2018  
2715           NaN  2016  
2254           NaN  2015

In [14]:
# filling missing values with 0
movie_gross["foreign_gross"] = movie_gross["foreign_gross"].fillna(0)

# Removing characters 
movie_gross["foreign_gross"].replace(",", "", inplace=True, regex=True)

# changinging datatype to float
movie_gross["foreign_gross"] = movie_gross["foreign_gross"].astype(float)

# checking movie_gross info
movie_gross.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   3387 non-null   float64
 4   year            3387 non-null   int64  
dtypes: float64(2), int64(1), object(2)
memory usage: 132.4+ KB


In [15]:
# A function find proportion of missing value
def missing_values(data):
    # identify the total missing values per column
    # sort in order
    data = data.reset_index()
    miss = data.isnull().sum().sort_values(ascending = False)

    # calculate percentage of the missing values
    percentage_miss = (data.isnull().sum() / len(data)).sort_values(ascending = False)

    # store in a dataframe
    missing = pd.DataFrame({"Missing Values": miss, "Percentage": percentage_miss}).reset_index()

    # remove values that are missing
    missing.drop(missing[missing["Percentage"] == 0].index, inplace = True)

    return missing

In [16]:
missing_values(movie_gross)

index  Missing Values  Percentage
0  domestic_gross              28    0.008267
1          studio               5    0.001476

In [17]:
# domestic_gross has 28 missing values. I will fill these with 0
movie_gross["domestic_gross"] = movie_gross["domestic_gross"].fillna(0)
# studio has 5 missing values. I drop these.
movie_gross.dropna(inplace=True)
missing_values(movie_gross)

Empty DataFrame
Columns: [index, Missing Values, Percentage]
Index: []

In [18]:
missing_values(movie_info)

index  Missing Values  Percentage
0  runtime_minutes           31739    0.217176
1           genres            5408    0.037005
2   original_title              21    0.000144

In [19]:
# there is no meaniful analysis that can be done on missing records so I drop them
movie_info.dropna(inplace=True)
missing_values(movie_info)

Empty DataFrame
Columns: [index, Missing Values, Percentage]
Index: []

In [20]:
missing_values(movie_ratings)

Empty DataFrame
Columns: [index, Missing Values, Percentage]
Index: []

movie_gross title column has movie names with the year in brackets which is a different format to primary_title in movie_info.

In [21]:
# Removing (year) from movies
movie_gross['title'] = movie_gross['title'].str.replace(r' \(\d{4}\)', '')
movie_gross.head()

title studio  domestic_gross  \
0                                  Toy Story 3     BV     415000000.0   
1                          Alice in Wonderland     BV     334200000.0   
2  Harry Potter and the Deathly Hallows Part 1     WB     296000000.0   
3                                    Inception     WB     292600000.0   
4                          Shrek Forever After   P/DW     238700000.0   

   foreign_gross  year  
0    652000000.0  2010  
1    691300000.0  2010  
2    664300000.0  2010  
3    535700000.0  2010  
4    513900000.0  2010

In [22]:
# separating genres into separate columns

movie_genre = movie_info['genres'].str.split(',')

max_genres = movie_genre.apply(len).max()

genre_columns = [f'genre{i}' for i in range(1, max_genres + 1)]

genre_df = movie_genre.apply(lambda x: pd.Series(x + [None] * (max_genres - len(x))))

genre_df.columns = genre_columns

movie_info.drop('genres', axis=1, inplace=True)

movie_info = pd.concat([movie_info, genre_df], axis=1)

movie_info


tconst                                      primary_title  \
0       tt0063540                                          Sunghursh   
1       tt0066787                    One Day Before the Rainy Season   
2       tt0069049                         The Other Side of the Wind   
4       tt0100275                           The Wandering Soap Opera   
5       tt0111414                                        A Thin Life   
...           ...                                                ...   
146134  tt9916160                                         Drømmeland   
146135  tt9916170                                      The Rehearsal   
146136  tt9916186  Illenau - die Geschichte einer ehemaligen Heil...   
146137  tt9916190                                          Safeguard   
146139  tt9916538                                Kuambil Lagi Hatiku   

                                           original_title  start_year  \
0                                               Sunghursh        2013   
1                                         Ashad Ka Ek Din        2019   
2                              The Other Side of the Wind        2018   
4                                   La Telenovela Errante        2017   
5                                             A Thin Life        2018   
...                                                   ...         ...   
146134                                         Drømmeland        2019   
146135                                           O Ensaio        2019   
146136  Illenau - die Geschichte einer ehemaligen Heil...        2017   
146137                                          Safeguard        2019   
146139                                Kuambil Lagi Hatiku        2019   

        runtime_minutes       genre1    genre2   genre3  
0                 175.0       Action     Crime    Drama  
1                 114.0    Biography     Drama     None  
2                 122.0        Drama      None     None  
4                  80.0       Comedy     Drama  Fantasy  
5                  75.0       Comedy      None     None  
...                 ...          ...       ...      ...  
146134             72.0  Documentary      None     None  
146135             51.0        Drama      None     None  
146136             84.0  Documentary      None     None  
146137             90.0        Drama  Thriller     None  
146139            123.0        Drama      None     None  

[112232 rows x 8 columns]

In [40]:
movie_info['genre2'].nunique()

26

### 2.2.2) Merging Datasets

In [23]:
# Merging movie_info and movie_ratings on "tconst"
movie_info_ratings_df  = pd.merge(movie_info, 
                              movie_ratings, 
                              on="tconst",
                              how="inner")

In [24]:
movie_info_ratings_df.head(2)

tconst                    primary_title   original_title  start_year  \
0  tt0063540                        Sunghursh        Sunghursh        2013   
1  tt0066787  One Day Before the Rainy Season  Ashad Ka Ek Din        2019   

   runtime_minutes     genre1 genre2 genre3  averagerating  numvotes  
0            175.0     Action  Crime  Drama            7.0        77  
1            114.0  Biography  Drama   None            7.2        43

In [25]:
movie_info_ratings_df.shape

(65720, 10)

In [26]:
movie_gross.shape

(3382, 5)

In [27]:
# Merging movie_info_ratings_df and movie_gross left_on "original_title" and right_on "title"
microsoft_studio_df  = pd.merge(movie_info_ratings_df, 
                              movie_gross, 
                              left_on=["primary_title","start_year"],
                              right_on=["title","year"],
                              how="inner")

In [28]:
microsoft_studio_df = microsoft_studio_df.drop(["year", "title"], axis=1)

In [29]:
microsoft_studio_df.head(2)

tconst primary_title original_title  start_year  runtime_minutes  \
0  tt0315642         Wazir          Wazir        2016            103.0   
1  tt0337692   On the Road    On the Road        2012            124.0   

      genre1 genre2   genre3  averagerating  numvotes   studio  \
0     Action  Crime    Drama            7.1     15378  Relbig.   
1  Adventure  Drama  Romance            6.1     37886      IFC   

   domestic_gross  foreign_gross  
0       1100000.0            0.0  
1        744000.0      8000000.0

In [30]:
microsoft_studio_df.shape

(2026, 13)

In [31]:
microsoft_studio_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2026 entries, 0 to 2025
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   tconst           2026 non-null   object 
 1   primary_title    2026 non-null   object 
 2   original_title   2026 non-null   object 
 3   start_year       2026 non-null   int64  
 4   runtime_minutes  2026 non-null   float64
 5   genre1           2026 non-null   object 
 6   genre2           1774 non-null   object 
 7   genre3           1229 non-null   object 
 8   averagerating    2026 non-null   float64
 9   numvotes         2026 non-null   int64  
 10  studio           2026 non-null   object 
 11  domestic_gross   2026 non-null   float64
 12  foreign_gross    2026 non-null   float64
dtypes: float64(4), int64(2), object(7)
memory usage: 221.6+ KB


In [32]:
# checking the size difference between merged dataset and revenue dataset
missing_movies = movie_gross[~movie_gross["title"].isin(microsoft_studio_df["primary_title"])]
missing_movies

title      studio  \
2           Harry Potter and the Deathly Hallows Part 1          WB   
13                                          Tron Legacy          BV   
28                                         Knight & Day         Fox   
31    Percy Jackson & The Olympians: The Lightning T...         Fox   
38                                                  Red        Sum.   
...                                                 ...         ...   
3380                       Furious (Legend of Kolovrat)     CARUSEL   
3381                                Beauty and the Dogs       Osci.   
3383                        Edward II (2018 re-release)          FM   
3384                                           El Pacto        Sony   
3385                                           The Swan  Synergetic   

      domestic_gross  foreign_gross  year  
2        296000000.0    664300000.0  2010  
13       172100000.0    228000000.0  2010  
28        76400000.0    185500000.0  2010  
31        88800000.0    137700000.0  2010  
38        90400000.0    108600000.0  2010  
...              ...            ...   ...  
3380         10000.0            0.0  2018  
3381          8900.0            0.0  2018  
3383          4800.0            0.0  2018  
3384          2500.0            0.0  2018  
3385          2400.0            0.0  2018  

[1382 rows x 5 columns]

In [33]:
missing_movies[missing_movies["title"].str.contains(r'\(.*\)', regex=True)]

title   studio  \
63                       Aftershock (Tangshan Dadizhen)       CL   
79           If You Are the One 2 (Fei Cheng Wu Rao II)       CL   
150                             A Prophet (Un prophète)      SPC   
237             The Salvation Poem (Poema de Salvacion)       CZ   
247                                Hideaway (Le Refuge)   Strand   
...                                                 ...      ...   
3297                The Big Lebowski (20th Anniversary)   Fathom   
3306                         Vertigo (60th Anniversary)   Fathom   
3317  Maquia: When the Promised Flower Blooms (Subti...    Elev.   
3380                       Furious (Legend of Kolovrat)  CARUSEL   
3383                        Edward II (2018 re-release)       FM   

      domestic_gross  foreign_gross  year  
63           63000.0    100200000.0  2010  
79          427000.0     75600000.0  2010  
150        2100000.0     15800000.0  2010  
237         915000.0       183000.0  2010  
247          34500.0       837000.0  2010  
...              ...            ...   ...  
3297        583000.0            0.0  2018  
3306        394000.0            0.0  2018  
3317        186000.0            0.0  2018  
3380         10000.0            0.0  2018  
3383          4800.0            0.0  2018  

[91 rows x 5 columns]

### 2.2.3) Feature Engineering

Addding `total_revenue` column to analyze revenue

In [34]:
microsoft_studio_df["total_revenue"] = microsoft_studio_df["domestic_gross"] + \
microsoft_studio_df["foreign_gross"]
microsoft_studio_df.head(2)

tconst primary_title original_title  start_year  runtime_minutes  \
0  tt0315642         Wazir          Wazir        2016            103.0   
1  tt0337692   On the Road    On the Road        2012            124.0   

      genre1 genre2   genre3  averagerating  numvotes   studio  \
0     Action  Crime    Drama            7.1     15378  Relbig.   
1  Adventure  Drama  Romance            6.1     37886      IFC   

   domestic_gross  foreign_gross  total_revenue  
0       1100000.0            0.0      1100000.0  
1        744000.0      8000000.0      8744000.0